x/superfluid/keeper/intermediary_account.go

```diff
@@ -107,10 +110,76 @@ func (k Keeper) SetIntermediaryAccount(ctx sdk.Context, acc types.SuperfluidInte
 	prefixStore.Set(acc.GetAccAddress(), bz)
 }
 
-func (k Keeper) DeleteIntermediaryAccount(ctx sdk.Context, address sdk.AccAddress) {
+func (k Keeper) DeleteAllEmptyIntermediaryAccounts(ctx sdk.Context) {
+	itermedairyAccounts := k.GetAllIntermediaryAccounts(ctx)
+	for _, intermediaryAccount := range itermedairyAccounts {
+		k.DeleteIntermediaryAccountIfNoDelegation(ctx, intermediaryAccount)
+	}
+}
+
+// DeleteIntermediaryAccount deletes given intermediary account from store
+// Note that intermediary account is highly related to staking and delgation, and this
+// method should be used with caution.
+func (k Keeper) DeleteIntermediaryAccountIfNoDelegation(ctx sdk.Context, intermedairyAcc types.SuperfluidIntermediaryAccount) {
+	// check if any delegations or intermediary connections exist
+	if k.IntermediaryAccountDelegationsExists(ctx, intermedairyAcc) {
+		return
+	}
+
 	store := ctx.KVStore(k.storeKey)
+
+	// store for intermediary account
 	prefixStore := prefix.NewStore(store, types.KeyPrefixIntermediaryAccount)
-	prefixStore.Delete(address)
+	prefixStore.Delete(intermedairyAcc.GetAccAddress())
+}
+
+// IntermediaryAccountDelegationsExists returns true if the gien intermediary account has any delegations.
+// We check this by
+// - using staking keeper to check if actual delegations exist,
+// - checking if there are any intermediary account connection remaining to the intermediaryAcc
+// - if there's no synthetic lock with the intermediaryAccount.Denom + intermediaryAccount.ValAddr combination
```

Comment:
nit: some of the points were read as "check by checking"
```suggestion
// We determine this by
// - using staking keeper to check if actual delegations exist,
// - checking if there are any intermediary account connection remaining to the intermediaryAcc
// - checking if there's no synthetic lock with the intermediaryAccount.Denom + intermediaryAccount.ValAddr combination
```